Refactor Server Input Handling

sjrc6 opened the pull request:
<!-- What is the motivation for the changes of this pull request? -->
This is version 2 of this PR description, the old one should be in the edit history.

## New changes to this PR at of Jun 11th, 2025
- Compressed the commits from 16 to 5
- The function renaming and comment documentation of existing behavior not affected by this PR has been almost entirely removed. It can be done in a future PR instead. This should make it easier to understand and discuss the changes.
- Cleaned up some of the code, SetEvalTick/SetStartTick is now a less privileged EarlyEvalTick()/EarlyStartTick()
- Instead of moving the `m_NumInputs++;` in character OnDirectInput to OnPredictedInput we keep it in OnDirectInput but make sure that OnDirectInput only runs when there is an actual input to process. (which is exactly how it worked before) Both ways produce correct physics but this way seems safer.
- [This commit](https://github.com/ddnet/ddnet/pull/8432/commits/41f73c34d13ffe9d5866e0cb3243e0ff0e2d2c85) has some more notes in the description about the DirectInput changes.

## Summary

The first part of this change allows the server to handle out of order and duplicate inputs by changing the unordered duplicates-allowed input buffer to a strict order buffer with predictable positions for the input of each gametick.

The implemented behavior is to disallow the client from changing inputs for a specific gametick in the server input buffer after it has been sent. This could go either way imo, there might be some small reliability benefits to allowing the client to change the inputs afterwards but it also opens some new possibilities for malicious clients to "change their mind" about what they want which would make some types of bots stronger. Ensuring the input buffer never changes also makes #10329 or similar features more stable as they can trust that the inputs won't change.

Because we aren't allowing multiple inputs on the same tick anymore this makes a lot of the previous code seem redundant/duplicated unless you've seen this PR. 

So the second part of this change moves the handling of 'DirectInput' physics behavior from the server.cpp 'CServer::Run()' to 'CGameWorld::Tick()', and removes the 'OnClientPredictedEarlyInput' loop from 'CServer::Run()' as 1 input per tick makes it redundant. 

The tick lifecycle therefore changes

Previous behavior:

```
CServer::Run(){
  OnClientPredictedEarlyInput() -> OnPredictedEarlyInput() -> CCharacter::DirectInput() -> FireWeapon() (spawn projectiles)
  GameTick++
  OnClientPredictedInput()
  GameTick() -> Character Tick -> HandleWeapons() -> FireWeapon() (also possibly spawn projectiles)
}
```

8432 behavior:
```
CServer::Run(){
  GameTick++
  OnClientPredictedInput()
  GameTick() -> 
  {
   CCharacter::DirectInput() -> FireWeapon(EarlyTick = true) (spawn projectiles, with start ticks as if they were from the previous gametick)
   Character Tick -> HandleWeapons() -> FireWeapon(EarlyTick = false) (also possibly spawn projectiles)
  }
}
```

This makes FireWeapon() slightly more ugly than it was before because it now has to handle the new behavior of spawning entities with a with a different start tick, but I think this directly obvious behavior is better than the previous obscured and confusing structure which had the same result. I'm open to suggests to change the way the projectile starting ticks are set in FireWeapon() if there are other opinions.



## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [x] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

sjrc6 commented:
It's possible that m_LatestInput and m_Input can stop being separate variables but that looks like too much work for only risk so I didn't touch it.

trml commented:
In #1441 it was changed so that the order depended on client ID. Before that it was dependent on the order that the inputs arrived at the server in, and also asynchronously from IntendedTick (this is how it originally was and still is in vanilla).
Then after #1922 weapon code was changed so that hammer/laser unfreeze (like hammering a frozen tee that hammers you back) would still work regardless of the order of direct inputs (or the order of IDs).
 
With this change the order will be the same as the Tick() order. However since it was already changed once before (kind of unintentionally), and since the weapon physics now depends less on the order of the inputs, I don't think it would be a problem to change it again.

Unless there are some tricks that require a very specific order of firing weapons that will no longer be possible once the order of the two are the same, but in that case it could be argued the that the solution would be to either change physics again or remove the tricks, rather than keeping the current ID-dependent code.

sjrc6 commented:
If its not that hard I think it would be good to preserve the client order physics. Just for the sake of consistency.

heinrich5991 commented:
Modifying input handling is very dangerous and can lead to unintended physics handling. The reviewer should be very careful. Also, this will need to be reviewed for interaction with teehistorian before it is merged.

sjrc6 commented:
> Modifying input handling is very dangerous and can lead to unintended physics handling. The reviewer should be very careful. Also, this will need to be reviewed for interaction with teehistorian before it is merged.

I'm well aware of the risks. 

Addressing your teehistorian concern, the only change is moving

```
if(m_TeeHistorianActive)
{
    m_TeeHistorian.RecordPlayerInput(ClientId, m_apPlayers[ClientId]->GetUniqueCid(), pApplyInput);
}
```

From `OnClientPredictedEarlyInput` to `OnClientPredictedInput`

The main difference between these two functions was that OnClientPredictedInput is only called on the earliest input which is intended for the current tick tick that was sent by the client

 (except for a semi-rare edge case bug where the client sends an input at position 199 and then another input in the same tick which then rolls the buffer over to 0, then the loop incorrectly stops at position 0 before reaching the earlier input at the end of the buffer, but this never happened because directinput didn't exist)

`OnClientPredictedEarlyInput`:
https://github.com/ddnet/ddnet/blob/64f4ea1512b5abcc1ef27b90ede05cb9934df69b/src/engine/server/server.cpp#L2910-L2917

`OnClientPredictedInput`:
https://github.com/ddnet/ddnet/blob/64f4ea1512b5abcc1ef27b90ede05cb9934df69b/src/engine/server/server.cpp#L2931-L2939

(These actually check for the same gametick because the `m_CurrentGameTick` is incremented between them)

The call to `m_TeeHistorian.RecordPlayerInput() `was previously placed into `OnClientPredictedEarlyInput` which means it would have received multiple inputs on the same tick if the client had sent any.

But because there was only ever 1 input put tick being sent by clients we don't need 2 different versions of this function. The function which was capable of being called multiple times (`OnClientPredictedEarlyInput` ) is removed by this PR so I simply put `m_TeeHistorian.RecordPlayerInput() ` into `OnClientPredictedInput ` since it will always receive all inputs now that 1 input per tick is enforced by the server.

JSaurusRex commented:
> Additionally somewhat related to this PR if https://github.com/ddnet/ddnet/pull/8372 is merged and client input re-sending is implementing it would in theory allow 2 (or more) cooperating clients to play on to up the full ~2000ms prediction margin without experiencing additional latency and having an entire 2 seconds of protection from packet loss.

Only problem with that is that as the PR is now any missed inputs won't be send to the other client, the client needs all the inputs done to predict accurately. So a missed input wouldn't work with the current model, even out of order inputs wouldn't work

sjrc6 commented:
> Only problem with that is that as the PR is now any missed inputs won't be send to the other client, the client needs all the inputs done to predict accurately. So a missed input wouldn't work with the current model, even out of order inputs wouldn't work

Functionality of #8372 with network instability can be a separate concern. Even without this PR #8372 would fail with inputs arriving out of order because the server does still process them just incorrectly.

sjrc6 commented:
This should be ready to review, I've gone over it enough times that I've run out of ideas for how it could break things. 

I probably won't have time to finish #8441 for the next few weeks because it ended up being more difficult than expected to get a reasonable bandwidth. So there's no rush to merge this right away but it is "finished" notwithstanding any requested changes.

fokkonaut commented on src/game/server/entities/character.cpp:
```diff
@@ -546,7 +546,7 @@ void CCharacter::FireWeapon()
 			int Lifetime = (int)(Server()->TickSpeed() * GetTuning(m_TuneZone)->m_GunLifetime);
 
-			new CProjectile(
+			CProjectile *Projectile = new CProjectile(
```
pProjectile

sjrc6 committed:
Change servers unordered input buffer to fixed position

sjrc6 committed:
Prevent 'm_LastAckedSnapshot' from going backwards if inputs arrive late

sjrc6 committed:
Remove unnecessary 'm_LatestInput' member from server

sjrc6 committed:
Move direct input functionality to within the gametick

NOTES:
- The physics is not correct at this commit, it should be corrected in the following commit
- The Gameworld now takes the responsibility of handling 'DirectInput' behavior
- 'CPlayer::OnPredictedInput' Takes the responsibility of updating 'm_PlayerFlags' and 'm_Spawning' as 'CPlayer::OnPredictedEarlyInput' no longer exists. Because there is only ever 1 input per tick this is exactly the same as it would have been if it happened in 'CPlayer::OnPredictedEarlyInput'
- DirectInput no longer takes an input as a parameter, because there is only 1 input per tick we can always use m_Input which is set ahead of time by the 'CGameContext::OnClientPredictedInput' -> 'CPlayer::OnPredictedInput' -> 'CCharacter::OnPredictedInput' chain
- Because 'CCharacter::OnDirectInput()' (currently a bad name) is called as part of the gametick we only want to do it when there was an actual input for that tick. So we track that.

sjrc6 committed:
Correct physics for new DirectInput behavior

SollyBunny reviewed (approved):
plz merg

Robyt3 reviewed (commented):
I'm afraid of changing physics especially with the last two commits. Is it possible to split the first three commits into a separate PR?

`TEEHISTORIAN_VERSION_MINOR` should also be incremented.

sjrc6 committed:
bump TEEHISTORIAN_VERSION_MINOR

sjrc6 commented:
> I'm afraid of changing physics especially with the last two commits. Is it possible to split the first three commits into a separate PR?

I've tested it with TAS playback tools, is there anything in particular you're afraid of breaking? I think the code is much cleaner this way.

Robyt3 commented:
> I've tested it with TAS playback tools, is there anything in particular you're afraid of breaking?

Mostly 41f73c34d13ffe9d5866e0cb3243e0ff0e2d2c85 and 00fbda6bca6eadaad894dad58895abafc4b9cd94 because they explicitly change the input behavior back and forth. They should rather be squashed since it doesn't make sense to have a state where only one of them is merged.

> The CI failed for no reason

We probably need to bump some timeout values.

sjrc6 commented:
> Mostly [41f73c3](https://github.com/ddnet/ddnet/commit/41f73c34d13ffe9d5866e0cb3243e0ff0e2d2c85) and [00fbda6](https://github.com/ddnet/ddnet/commit/00fbda6bca6eadaad894dad58895abafc4b9cd94) because they explicitly change the input behavior back and forth. They should rather be squashed since it doesn't make sense to have a state where only one of them is merged.

I did it that to make it easier to review, that way you can see which exactly which part of the physics change required to make it correct. 

I don't think the expectation is that every commit with a PR is valid for production use? It still compiles fine and the physics bug is very predictable, projectiles just have a lifetime that is 1 tick too short because the gametick when they are spawned is increased by 1. So you can see in the following commit it is fixed by giving them the correct lifetime, that way it has the same behavior as before, this makes it easier to test too.

I can squash those if you want, it makes no difference.

sjrc6 committed:
Merge branch 'master' into pr_input_refactor

heinrich5991 reviewed (commented):
I vote no on merging this unless a maintainer says they understand the code and there are no physics changes.

edg-l commented:
<details>
<summary>claude review</summary>

```
blocking      negative input tick reads out of bounds before it is validated   server.cpp:1804
blocking      direct weapon select fires one tick earlier, and mispredicts      character.cpp:415
question      resending an already-sent tick also drops its snapshot ack        server.cpp:1807
worth fixing  a burst of late inputs keeps the oldest and drops the newest      server.cpp:1844
```

**blocking — negative input tick reads out of bounds before it is validated**

The new fixed-position buffer indexes `m_aInputs` with `IntendedTick % 200` and dereferences it *before* the tick is range-checked.

https://github.com/ddnet/ddnet/blob/17ebe324c70c0b9a19870850acc85d180d098357/src/engine/server/server.cpp#L1804-L1815

`IntendedTick` is a raw `Unpacker.GetInt()` off the wire, and there is no state gate on the `NETMSG_INPUT` branch. C++ `%` truncates toward zero, so a client sending `IntendedTick = -199` gives `BufferPosition = -199` and `m_aInputs[-199].m_GameTick` reads ~100 KB before the heap-allocated `CServer`. On master the `IntendedTick < MIN_TICK` guard at line 1812 rejected this before any indexing. Move the duplicate-input block below that guard (after it, `IntendedTick` is in `[0, MAX_TICK)`, so `% 200` is always `0..199`); the second identical check at line 1849 is already correctly placed.

<details>
<summary>reproduced against a Release build</summary>

`sizeof(CInput)=516`, `offsetof(CServer,m_aClients)+offsetof(CClient,m_aInputs)=264`, `CServer` is `new CServer()` (25 MB, its own mmap arena). Reading `m_aInputs[-199].m_GameTick` lands 101908 bytes before the object, in the unmapped page before the mapping:

```
&m_aInputs[-199].tick  = 0x7fd68a0ad1fc  (101908 bytes before base)
SIGSEGV on the read  -> remote server crash (DoS)
```

Crashes on every run. Moving the block below line 1812 builds clean and closes it.

</details>

---

**blocking — direct weapon select fires one tick earlier, and the client mispredicts it**

This is the physics change. Not in the firing order across clients (that is preserved, `m_apPlayers` iteration is client-id order like the old loop) and not in projectile/laser start ticks (those are correctly back-dated). It is one line in `HandleWeaponSwitch`:

https://github.com/ddnet/ddnet/blob/17ebe324c70c0b9a19870850acc85d180d098357/src/game/server/entities/character.cpp#L414-L415

The gate reads `m_LatestInput` but the value reads `m_Input`. Moving DirectInput into `CGameWorld::Tick()` means `OnPredictedInput` (which writes `m_Input`) now runs *before* the fire instead of after it, so `m_Input` here holds this tick's input where it used to hold the previous tick's. `m_WantedWeapon` is persistent client state, so the two differ exactly on the tick a weapon key is pressed. The Next/Prev selection above is unaffected; it reads `m_LatestInput`/`m_LatestPrevInput`.

The client predicts the old behavior: `src/game/client/prediction/entities/character.cpp:239-240` is the same line, and `gameclient.cpp:2395-2433` runs `OnDirectInput` before `OnPredictedInput`. So every direct weapon switch now mispredicts by one tick. Either snapshot `m_Input` at the top of `OnPredictedInput` and read that snapshot at line 415 to preserve behavior, or accept it as a deliberate physics change, mirror it in the prediction path, and say so (it can't land under "changed no physics").

<details>
<summary>the differential, one concrete tick</summary>

Player holding fire on grenade, reload expires at tick N, presses `+weapon5` so `input(N-1).m_WantedWeapon == 4`, `input(N).m_WantedWeapon == 5`. Gate `m_LatestInput.m_WantedWeapon == input(N) == 5` (true) in both.

```
                value read at :415         WantedWeapon      fires at tick N
master   m_Input == input(N-1) == 4        4-1 = grenade     grenade (switch at N+1)
head     m_Input == input(N)   == 5        5-1 = laser       laser
```

`m_WantedWeapon` is a level set on key-down in `CControls::ConKeyInputSet` and never cleared except by `+next`/`+prevweapon`, so `input(N-1)` and `input(N)` are equal on every tick except the press tick, where head switches one tick sooner than master and one tick sooner than the client still predicts.

</details>

---

**question — resending an already-sent tick now also drops that packet's snapshot ack and latency sample**

This is the concrete cost of disallowing input changes for an already-sent tick, which you flagged yourself. When the client fails to find a delta base or hits the CRC-error limit (`client.cpp:1988`, `2031`) it sets `m_aAckGameTick = -1` and calls `SendInput()` again *without advancing the pred tick*, so the resend carries the same `IntendedTick` but `LastAckedSnapshot = -1`. The duplicate check at server.cpp:1807 returns before line 1817, so the `-1` resync request, the latency sample and the `SNAPRATE_FULL` promotion are all skipped; recovery waits for the next pred tick. The `|| LastAckedSnapshot < 0` arm you added at line 1818 exists for exactly this `-1`, but the early return keeps it from being reached on the first attempt. Processing `LastAckedSnapshot`/latency before the duplicate-input return (drop only the input, not the ack) keeps it.

---

**worth fixing — a burst of late inputs keeps the oldest and drops the newest**

`PumpNetwork` drains queued packets with `Tick()` constant, so after a spike a client's inputs for ticks 100-103 all arrive while the server is at 102. `input(100)` is clamped by line 1844 to 103 and stamps slot `103`; 101, 102 and then the genuine `input(103)` all hit the duplicate check and are dropped. The one input that survives is the oldest, and the player's real tick-103 input never reaches the character. It's an edge case under packet loss and the direction is defensible versus master's multi-fire, but it's a second observable behavior change and belongs in the description rather than falling out of the clamp. The `TODO` you left at line 1842 points at the real fix: discard already-passed inputs instead of re-aiming them at the next tick.

---

minor: `FireWeapon(true)` back-dates `m_StartTick`/`m_EvalTick`/`m_AttackTick` per call site, and the hammer branch's `TakeDamage` → `SetEmote(EMOTE_PAIN, Server()->Tick()+25)` (`character.cpp:1034`) is not decremented, so a hammer-hit pain emote lasts one tick longer (cosmetic; it shows the `bool EarlyTick` + N decrements design is fragile, passing one intended tick down would be one value to keep right). `gameworld.cpp:221` uses `GetCharacter()` (null unless alive) where the old path used `m_pCharacter`, narrowing the DirectInput set for a character killed via `Die()` without `KillCharacter`; the new behavior is saner but it's a third behavior change in a PR described as a refactor.

Exercised: built server at base and head, verified the OOB crash with the offsets above and confirmed the reorder fix builds. The weapon-select divergence and the mispredict I traced through both server versions and the client prediction path by reading; I did not drive two servers with a scripted client to diff a teehistorian trace. Findings 3 and 4 I confirmed reachable from the client resend path but did not reproduce live.

</details>

edg-l commented:
reopening on request from author

heinrich5991 commented:
@edg-l is the maintainer who wants to review this.

edg-l commented:
i'll review myself once its finished again
